src/main/java/org/commcare/formplayer/session/FormplayerInstanceInitializer.java

```diff
@@ -61,10 +63,22 @@ protected InstanceRoot setupSessionData(ExternalDataInstance instance) {
 
         Hashtable<String, String> userProperties = u.getProperties();
 
+        String appLang = locale;
+        String[] locales = Localization.getGlobalLocalizerAdvanced().getAvailableLocales();
+        if (appLang == null || !Arrays.asList(locales).contains(appLang)) {
+            // the ordering is always ['default', <true default slug>, <all other slugs in order they were added>]
+            if (locales.length >= 2) {
+                appLang =  locales[1];
+            } else {
+                // to pass tests
+                appLang = locales[0];
+            }
+        }
```

Comment:
My read of the code here is you are trying to get the current selected locale ? If so, think we can use `Localization.getCurrentLocale()` directly.

Reply:
I wish I had known about this function sooner because this is doing exactly what I was looking to do from the start, without having to do all the unnecessary steps of passing the locale to the root generator function... I might end up reverting the commcare-core PR :(

Reply:
[partial revert PR](https://github.com/dimagi/commcare-core/pull/1447)